libpromises/evalfunction.c

```diff
@@ -8451,6 +8452,231 @@ static FnCallResult FnCallNetworkConnections(EvalContext *ctx, ARG_UNUSED const
+                GetErrorStr());
+            return FnReturnContext(false);
+        }
+        else if (read(fd, buf, sizeof(buf)) < 0)
+        {
+            Log(LOG_LEVEL_DEBUG, "Failed to read from file '%s': %s", path,
+                GetErrorStr());
+            close(fd);
+            return FnReturnContext(false);
+        }
+
+        close(fd);
+        return FnReturnContext(true);
+    }
+
+    // Else try read in separate thread, possibly blocking for N seconds
+    Log(LOG_LEVEL_DEBUG,
+        "Checking if file '%s' is readable in separate thread, "
+        "possibly blocking for %ld seconds.", path, timeout);
+
+    struct IsReadableThreadData thread_data = {0};
+    thread_data.path = path;
+
+    int ret = pthread_mutex_init(&thread_data.mutex, NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to initialize mutex: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    ret = pthread_cond_init(&thread_data.cond, NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to initialize condition: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    // Keep thread from doing its thing until we call
+    // pthread_cond_timedwait(3)
+    ret = pthread_mutex_lock(&thread_data.mutex);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to lock mutex: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    ret = pthread_create(&thread_data.thread, NULL, &IsReadableThreadRoutine,
+                         &thread_data);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to create thread: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    FnCallResult result;
+    // Wait on thread to finish or timeout
+    ret = ThreadWait(&thread_data.cond, &thread_data.mutex, timeout);
+    switch (ret)
+    {
+        case 0: // Thread finished in time
+            result = thread_data.result;
+            break;
+
+        case ETIMEDOUT: // Thread timed out
+            Log(LOG_LEVEL_DEBUG, "File '%s' is not readable: "
+                "Read operation timed out, exceeded %ld seconds.", path,
+                timeout);
+
+            ret = pthread_cancel(thread_data.thread);
+            if (ret != 0)
+            {
+                Log(LOG_LEVEL_ERR, "Failed to cancel thread");
+                return FnFailure();
+            }
+
+            result = FnReturnContext(false);
+            break;
+
+        default:
+            Log(LOG_LEVEL_ERR, "Failed to wait for condition: %s",
+                GetErrorStrFromCode(ret));
+            return FnFailure();
+
+    }
+
+    ret = pthread_mutex_unlock(&thread_data.mutex);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to unlock mutex");
+        return FnFailure();
+    }
+
+    void *status;
+    ret = pthread_join(thread_data.thread, &status);
```

Comment:
What about using `pthread_tryjoin_np()` on platforms that have it?

Reply:
If thread does not have time to exit before we call `pthread_tryjoin_np()` it will leave us with a zombie thread.

Reply:
You'd have to detach the thread in such cases.

Reply:
Yes, absolutely.